File: app/src/main/java/com/byd/dashcast/ui/main/PermissionBannerCoordinator.java
package com.byd.dashcast.ui.main;

import android.content.Context;
import android.view.View;

import com.byd.dashcast.AdbLocalClient;
import com.byd.dashcast.AppLogger;
import com.byd.dashcast.R;
import com.byd.dashcast.data.prefs.ClusterPrefs;
import com.byd.dashcast.ime.ClusterImeWatcherService;
import com.byd.dashcast.platform.Platform;
import com.byd.dashcast.proxy.ShellGateway;

/**
 * Owns the two onboarding banners:
 * <ul>
 *   <li>IME / Accessibility service banner (DL5 only) — {@code card_ime_a11y_banner}
 *   <li>HUD notification-listener banner (all devices) — {@code card_hud_notif_banner}
 * </ul>
 * Call {@link #refresh()} from {@code onStart} / {@code onResume} to recompute visibility.
 */
public final class PermissionBannerCoordinator {

    private static final String TAG = "PermBannerCoord";

    public interface Host {
        Context getContext();
        void runOnUiThread(Runnable r);
        /** Returns false once the host Activity is finishing or destroyed. */
        boolean isActivityAlive();
        void startActivity(android.content.Intent intent);
    }

    private final View mCardIme;
    private final View mCardHudNotif;
    private final Host mHost;

    public PermissionBannerCoordinator(View cardIme, View cardHudNotif, Host host) {
        mCardIme      = cardIme;
        mCardHudNotif = cardHudNotif;
        mHost         = host;
        try { setupIme();      } catch (Throwable t) { AppLogger.e(TAG, "setupIme failed", t); }
        try { setupHudNotif(); } catch (Throwable t) { AppLogger.e(TAG, "setupHudNotif failed", t); }
    }

    /** Recomputes both banners' visibility. Safe to call repeatedly from lifecycle methods. */
    public void refresh() {
        try { refreshIme();      } catch (Throwable t) { AppLogger.e(TAG, "refreshIme failed", t); }
        try { refreshHudNotif(); } catch (Throwable t) { AppLogger.e(TAG, "refreshHudNotif failed", t); }
    }

    // ── IME / A11Y banner ─────────────────────────────────────────────────────

    private void setupIme() {
        if (mCardIme == null) return;
        View btnEnable  = mCardIme.findViewById(R.id.btn_ime_banner_enable);
        View btnLater   = mCardIme.findViewById(R.id.btn_ime_banner_later);
        View btnDismiss = mCardIme.findViewById(R.id.btn_ime_banner_dismiss);

        if (btnEnable  != null) btnEnable.setOnClickListener(v -> enableImeA11yOneClick(v));
        if (btnLater   != null) btnLater.setOnClickListener(v -> mCardIme.setVisibility(View.GONE));
        if (btnDismiss != null) {
            btnDismiss.setOnClickListener(v -> {
                try { ClusterPrefs.setImeBannerDismissed(mHost.getContext()); } catch (Throwable ignored) {}
                mCardIme.setVisibility(View.GONE);
            });
        }
        refreshIme();
    }

    private void refreshIme() {
        if (mCardIme == null) return;
        boolean shouldShow = false;
        try {
            if (Platform.get().isDiLink5(mHost.getContext())) {
                boolean dismissed = ClusterPrefs.isImeBannerDismissed(mHost.getContext());
                boolean enabled   = ClusterImeWatcherService.isEnabled(mHost.getContext());
                shouldShow = !dismissed && !enabled;
                if (enabled && mCardIme.getVisibility() == View.VISIBLE) {
                    try {
                        android.widget.Toast.makeText(mHost.getContext(),
                                R.string.ime_banner_toast_enabled,
                                android.widget.Toast.LENGTH_SHORT).show();
                    } catch (Throwable ignored) {}
                }
            }
        } catch (Throwable t) {
            AppLogger.e(TAG, "refreshIme inner failed", t);
        }
        mCardIme.setVisibility(shouldShow ? View.VISIBLE : View.GONE);
    }

    private void enableImeA11yOneClick(final View btnEnable) {
        if (btnEnable != null) btnEnable.setEnabled(false);
        try {
            final String comp = "com.byd.dashcast/com.byd.dashcast.ime.ClusterImeWatcherService";
            final String cmd =
                    "COMP='" + comp + "'; "
                  + "CURRENT=$(settings get secure enabled_accessibility_services 2>/dev/null); "
                  + "if [ \"$CURRENT\" = \"null\" ] || [ -z \"$CURRENT\" ]; then "
                  +   "NEW=\"$COMP\"; "
                  + "elif echo \"$CURRENT\" | grep -q \"$COMP\"; then "
                  +   "NEW=\"$CURRENT\"; "
                  + "else "
                  +   "NEW=\"$CURRENT:$COMP\"; "
                  + "fi; "
                  + "settings put secure enabled_accessibility_services \"$NEW\"; "
                  + "settings put secure accessibility_enabled 1; "
                  + "echo OUT=$(settings get secure enabled_accessibility_services)";

            ShellGateway.execShellWithResult(mHost.getContext(), cmd, new AdbLocalClient.Callback() {
                @Override public void onSuccess(final String report) {
                    mHost.runOnUiThread(() -> {
                        if (!mHost.isActivityAlive()) return;
                        boolean ok = ClusterImeWatcherService.isEnabled(mHost.getContext());
                        if (ok) {
                            AppLogger.i(TAG, "IME a11y enabled via shell ✓");
                            try {
                                android.widget.Toast.makeText(mHost.getContext(),
                                        R.string.ime_banner_toast_enabled,
                                        android.widget.Toast.LENGTH_SHORT).show();
                            } catch (Throwable ignored) {}
                            if (mCardIme != null) mCardIme.setVisibility(View.GONE);
                            if (btnEnable != null) btnEnable.setEnabled(true);
                        } else {
                            AppLogger.w(TAG, "shell OK but a11y not enabled — fallback. report=" + report);
                            openA11ySettingsFallback();
                            if (btnEnable != null) btnEnable.setEnabled(true);
                        }
                    });
                }
                @Override public void onError(final String error) {
                    mHost.runOnUiThread(() -> {
                        if (!mHost.isActivityAlive()) return;
                        AppLogger.w(TAG, "one-click a11y shell failed: " + error + " — fallback");
                        openA11ySettingsFallback();
                        if (btnEnable != null) btnEnable.setEnabled(true);
                    });
                }
            });
        } catch (Throwable t) {
            AppLogger.e(TAG, "enableImeA11yOneClick threw", t);
            openA11ySettingsFallback();
            if (btnEnable != null) btnEnable.setEnabled(true);
        }
    }

    private void openA11ySettingsFallback() {
        try {
            android.content.Intent i = new android.content.Intent(
                    android.provider.Settings.ACTION_ACCESSIBILITY_SETTINGS);
            i.addFlags(android.content.Intent.FLAG_ACTIVITY_NEW_TASK);
            mHost.startActivity(i);
            return;
        } catch (Throwable t) {
            AppLogger.w(TAG, "ACTION_ACCESSIBILITY_SETTINGS unavailable: " + t.getMessage());
        }
        try {
            android.content.Intent i = new android.content.Intent();
            i.setComponent(new android.content.ComponentName(
                    "com.android.settings",
                    "com.android.settings.Settings$AccessibilitySettingsActivity"));
            i.addFlags(android.content.Intent.FLAG_ACTIVITY_NEW_TASK);
            mHost.startActivity(i);
            return;
        } catch (Throwable t) {
            AppLogger.w(TAG, "direct AccessibilitySettingsActivity unavailable: " + t.getMessage());
        }
        try {
            android.content.Intent i = new android.content.Intent(
                    android.provider.Settings.ACTION_SETTINGS);
            i.addFlags(android.content.Intent.FLAG_ACTIVITY_NEW_TASK);
            mHost.startActivity(i);
        } catch (Throwable t) {
            AppLogger.e(TAG, "no Settings activity reachable on this ROM", t);
            try {
                android.widget.Toast.makeText(mHost.getContext(),
                        R.string.ime_banner_toast_cannot_open_settings,
                        android.widget.Toast.LENGTH_LONG).show();
            } catch (Throwable ignored) {}
        }
    }

    // ── HUD notification-listener banner ──────────────────────────────────────

    private void setupHudNotif() {
        if (mCardHudNotif == null) return;
        View btnEnable  = mCardHudNotif.findViewById(R.id.btn_hud_banner_enable);
        View btnLater   = mCardHudNotif.findViewById(R.id.btn_hud_banner_later);
        View btnDismiss = mCardHudNotif.findViewById(R.id.btn_hud_banner_dismiss);

        if (btnEnable  != null) btnEnable.setOnClickListener(v -> enableHudNotifOneClick(v));
        if (btnLater   != null) btnLater.setOnClickListener(v -> mCardHudNotif.setVisibility(View.GONE));
        if (btnDismiss != null) {
            btnDismiss.setOnClickListener(v -> {
                try { ClusterPrefs.setHudBannerDismissed(mHost.getContext()); } catch (Throwable ignored) {}
                mCardHudNotif.setVisibility(View.GONE);
            });
        }
        refreshHudNotif();
    }

    private void refreshHudNotif() {
        if (mCardHudNotif == null) return;
        boolean shouldShow = false;
        try {
            boolean dismissed = ClusterPrefs.isHudBannerDismissed(mHost.getContext());
            if (!dismissed) {
                String listeners = android.provider.Settings.Secure.getString(
                        mHost.getContext().getContentResolver(), "enabled_notification_listeners");
                boolean granted = listeners != null && listeners.contains(mHost.getContext().getPackageName());
                shouldShow = !granted;
            }
        } catch (Throwable t) {
            AppLogger.e(TAG, "refreshHudNotif inner failed", t);
        }
        mCardHudNotif.setVisibility(shouldShow ? View.VISIBLE : View.GONE);
    }

    private void enableHudNotifOneClick(final View btnEnable) {
        if (btnEnable != null) btnEnable.setEnabled(false);
        try {
            final String comp = "com.byd.dashcast/com.byd.dashcast.hud.MapNotificationListenerService";
            final String cmd  = "cmd notification allow_listener " + comp;

            ShellGateway.execShellWithResult(mHost.getContext(), cmd, new AdbLocalClient.Callback() {
                @Override public void onSuccess(final String report) {
                    mHost.runOnUiThread(() -> {
                        if (!mHost.isActivityAlive()) return;
                        String listeners = android.provider.Settings.Secure.getString(
                                mHost.getContext().getContentResolver(), "enabled_notification_listeners");
                        boolean ok = listeners != null && listeners.contains(mHost.getContext().getPackageName());
                        if (ok) {
                            AppLogger.i(TAG, "Notification listener enabled via shell ✓");
                            try {
                                android.widget.Toast.makeText(mHost.getContext(),
                                        R.string.hud_notif_banner_toast_enabled,
                                        android.widget.Toast.LENGTH_SHORT).show();
                            } catch (Throwable ignored) {}
                            if (mCardHudNotif != null) mCardHudNotif.setVisibility(View.GONE);
                            if (btnEnable != null) btnEnable.setEnabled(true);
                        } else {
                            AppLogger.w(TAG, "shell OK but listener not enabled — fallback. report=" + report);
                            openNotifListenerSettingsFallback();
                            if (btnEnable != null) btnEnable.setEnabled(true);
                        }
                    });
                }
                @Override public void onError(final String error) {
                    mHost.runOnUiThread(() -> {
                        if (!mHost.isActivityAlive()) return;
                        AppLogger.w(TAG, "one-click notif listener shell failed: " + error + " — fallback");
                        openNotifListenerSettingsFallback();
                        if (btnEnable != null) btnEnable.setEnabled(true);
                    });
                }
            });
        } catch (Throwable t) {
            AppLogger.e(TAG, "enableHudNotifOneClick threw", t);
            openNotifListenerSettingsFallback();
            if (btnEnable != null) btnEnable.setEnabled(true);
        }
    }

    private void openNotifListenerSettingsFallback() {
        try {
            android.content.Intent i = new android.content.Intent(
                    android.provider.Settings.ACTION_NOTIFICATION_LISTENER_SETTINGS);
            i.addFlags(android.content.Intent.FLAG_ACTIVITY_NEW_TASK);
            mHost.startActivity(i);
        } catch (Throwable t) {
            AppLogger.w(TAG, "ACTION_NOTIFICATION_LISTENER_SETTINGS unavailable: " + t.getMessage());
        }
    }
}

File: app/src/main/java/com/byd/dashcast/ui/main/PermissionBannerCoordinator.kt
package com.byd.dashcast.ui.main

import android.content.ComponentName
import android.content.Context
import android.content.Intent
import android.provider.Settings
import android.view.View
import android.widget.Toast

import androidx.core.view.isVisible

import com.byd.dashcast.R
import com.byd.dashcast.data.prefs.ClusterPrefs
import com.byd.dashcast.ime.ClusterImeWatcherService
import com.byd.dashcast.infrastructure.AdbLocalClient
import com.byd.dashcast.platform.Platform
import com.byd.dashcast.proxy.ShellGateway
import com.byd.dashcast.util.AppLogger

/**
 * Owns the two onboarding banners:
 *  - IME / Accessibility service banner (DL5 only) — `card_ime_a11y_banner`
 *  - HUD notification-listener banner (all devices) — `card_hud_notif_banner`
 *
 * Call [refresh] from `onStart` / `onResume` to recompute visibility.
 *
 * Kotlin port note: every `$` in the accessibility shell script below is escaped as `\$`.
 * CURRENT, COMP and NEW are SHELL variables and all three are valid Kotlin identifiers, so an
 * unescaped port would interpolate them into nothing and silently ship a broken command. Same
 * for the `Settings$AccessibilitySettingsActivity` component name.
 */
class PermissionBannerCoordinator(
        private val mCardIme: View?,
        private val mCardHudNotif: View?,
        private val mHost: Host
) {

    interface Host {
        fun getContext(): Context
        // Named 'runnable' to agree with SplitController.Host: MainActivity implements BOTH
        // interfaces and satisfies this method by inheriting Activity.runOnUiThread, so the two
        // declarations must not disagree on the parameter name.
        fun runOnUiThread(runnable: Runnable)
        /** Returns false once the host Activity is finishing or destroyed. */
        fun isActivityAlive(): Boolean
        fun startActivity(intent: Intent)
    }

    init {
        try { setupIme() } catch (t: Throwable) { AppLogger.e(TAG, "setupIme failed", t) }
        try { setupHudNotif() } catch (t: Throwable) { AppLogger.e(TAG, "setupHudNotif failed", t) }
    }

    /** Recomputes both banners' visibility. Safe to call repeatedly from lifecycle methods. */
    fun refresh() {
        try { refreshIme() } catch (t: Throwable) { AppLogger.e(TAG, "refreshIme failed", t) }
        try { refreshHudNotif() } catch (t: Throwable) { AppLogger.e(TAG, "refreshHudNotif failed", t) }
    }

    // ── IME / A11Y banner ─────────────────────────────────────────────────────

    private fun setupIme() {
        val card = mCardIme ?: return
        val btnEnable: View? = card.findViewById(R.id.btn_ime_banner_enable)
        val btnLater: View? = card.findViewById(R.id.btn_ime_banner_later)
        val btnDismiss: View? = card.findViewById(R.id.btn_ime_banner_dismiss)

        btnEnable?.setOnClickListener { v -> enableImeA11yOneClick(v) }
        btnLater?.setOnClickListener { card.visibility = View.GONE }
        btnDismiss?.setOnClickListener {
            try { ClusterPrefs.setImeBannerDismissed(mHost.getContext()) } catch (ignored: Throwable) {}
            card.visibility = View.GONE
        }
        refreshIme()
    }

    private fun refreshIme() {
        val card = mCardIme ?: return
        var shouldShow = false
        try {
            if (Platform.get().isDiLink5(mHost.getContext())) {
                val dismissed = ClusterPrefs.isImeBannerDismissed(mHost.getContext())
                val enabled = ClusterImeWatcherService.isEnabled(mHost.getContext())
                shouldShow = !dismissed && !enabled
                if (enabled && card.isVisible) {
                    try {
                        Toast.makeText(mHost.getContext(), R.string.ime_banner_toast_enabled,
                                Toast.LENGTH_SHORT).show()
                    } catch (ignored: Throwable) {}
                }
            }
        } catch (t: Throwable) {
            AppLogger.e(TAG, "refreshIme inner failed", t)
        }
        card.visibility = if (shouldShow) View.VISIBLE else View.GONE
    }

    private fun enableImeA11yOneClick(btnEnable: View?) {
        btnEnable?.isEnabled = false
        try {
            val comp = "com.byd.dashcast/com.byd.dashcast.ime.ClusterImeWatcherService"
            // Every \$ below is a SHELL reference, not a Kotlin template.
            val cmd =
                    "COMP='" + comp + "'; " +
                    "CURRENT=\$(settings get secure enabled_accessibility_services 2>/dev/null); " +
                    "if [ \"\$CURRENT\" = \"null\" ] || [ -z \"\$CURRENT\" ]; then " +
                      "NEW=\"\$COMP\"; " +
                    "elif echo \"\$CURRENT\" | grep -q \"\$COMP\"; then " +
                      "NEW=\"\$CURRENT\"; " +
                    "else " +
                      "NEW=\"\$CURRENT:\$COMP\"; " +
                    "fi; " +
                    "settings put secure enabled_accessibility_services \"\$NEW\"; " +
                    "settings put secure accessibility_enabled 1; " +
                    "echo OUT=\$(settings get secure enabled_accessibility_services)"

            ShellGateway.execShellWithResult(mHost.getContext(), cmd, object : AdbLocalClient.Callback {
                // String? per the Batch 0 pin: AdbLocalClient.Callback is ported last and will
                // declare String?; a non-null override here would become an illegal narrowing.
                override fun onSuccess(out: String?) {
                    mHost.runOnUiThread {
                        if (!mHost.isActivityAlive()) return@runOnUiThread
                        val ok = ClusterImeWatcherService.isEnabled(mHost.getContext())
                        if (ok) {
                            AppLogger.i(TAG, "IME a11y enabled via shell ✓")
                            try {
                                Toast.makeText(mHost.getContext(),
                                        R.string.ime_banner_toast_enabled,
                                        Toast.LENGTH_SHORT).show()
                            } catch (ignored: Throwable) {}
                            mCardIme?.visibility = View.GONE
                            btnEnable?.isEnabled = true
                        } else {
                            AppLogger.w(TAG, "shell OK but a11y not enabled — fallback. report=" + out)
                            openA11ySettingsFallback()
                            btnEnable?.isEnabled = true
                        }
                    }
                }
                override fun onError(err: String?) {
                    mHost.runOnUiThread {
                        if (!mHost.isActivityAlive()) return@runOnUiThread
                        AppLogger.w(TAG, "one-click a11y shell failed: " + err + " — fallback")
                        openA11ySettingsFallback()
                        btnEnable?.isEnabled = true
                    }
                }
            })
        } catch (t: Throwable) {
            AppLogger.e(TAG, "enableImeA11yOneClick threw", t)
            openA11ySettingsFallback()
            btnEnable?.isEnabled = true
        }
    }

    private fun openA11ySettingsFallback() {
        try {
            val i = Intent(Settings.ACTION_ACCESSIBILITY_SETTINGS)
            i.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
            mHost.startActivity(i)
            return
        } catch (t: Throwable) {
            AppLogger.w(TAG, "ACTION_ACCESSIBILITY_SETTINGS unavailable: " + t.message)
        }
        try {
            val i = Intent()
            // \$ escape: this is a JVM nested-class binary name, not a Kotlin template.
            i.component = ComponentName("com.android.settings",
                    "com.android.settings.Settings\$AccessibilitySettingsActivity")
            i.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
            mHost.startActivity(i)
            return
        } catch (t: Throwable) {
            AppLogger.w(TAG, "direct AccessibilitySettingsActivity unavailable: " + t.message)
        }
        try {
            val i = Intent(Settings.ACTION_SETTINGS)
            i.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
            mHost.startActivity(i)
        } catch (t: Throwable) {
            AppLogger.e(TAG, "no Settings activity reachable on this ROM", t)
            try {
                Toast.makeText(mHost.getContext(),
                        R.string.ime_banner_toast_cannot_open_settings,
                        Toast.LENGTH_LONG).show()
            } catch (ignored: Throwable) {}
        }
    }

    // ── HUD notification-listener banner ──────────────────────────────────────

    private fun setupHudNotif() {
        val card = mCardHudNotif ?: return
        val btnEnable: View? = card.findViewById(R.id.btn_hud_banner_enable)
        val btnLater: View? = card.findViewById(R.id.btn_hud_banner_later)
        val btnDismiss: View? = card.findViewById(R.id.btn_hud_banner_dismiss)

        btnEnable?.setOnClickListener { v -> enableHudNotifOneClick(v) }
        btnLater?.setOnClickListener { card.visibility = View.GONE }
        btnDismiss?.setOnClickListener {
            try { ClusterPrefs.setHudBannerDismissed(mHost.getContext()) } catch (ignored: Throwable) {}
            card.visibility = View.GONE
        }
        refreshHudNotif()
    }

    private fun refreshHudNotif() {
        val card = mCardHudNotif ?: return
        var shouldShow = false
        try {
            val dismissed = ClusterPrefs.isHudBannerDismissed(mHost.getContext())
            if (!dismissed) {
                val listeners = Settings.Secure.getString(
                        mHost.getContext().contentResolver, "enabled_notification_listeners")
                val granted = listeners != null
                        && listeners.contains(mHost.getContext().packageName)
                shouldShow = !granted
            }
        } catch (t: Throwable) {
            AppLogger.e(TAG, "refreshHudNotif inner failed", t)
        }
        card.visibility = if (shouldShow) View.VISIBLE else View.GONE
    }

    private fun enableHudNotifOneClick(btnEnable: View?) {
        btnEnable?.isEnabled = false
        try {
            val comp = "com.byd.dashcast/com.byd.dashcast.hud.MapNotificationListenerService"
            val cmd = "cmd notification allow_listener " + comp

            ShellGateway.execShellWithResult(mHost.getContext(), cmd, object : AdbLocalClient.Callback {
                override fun onSuccess(out: String?) {
                    mHost.runOnUiThread {
                        if (!mHost.isActivityAlive()) return@runOnUiThread
                        val listeners = Settings.Secure.getString(
                                mHost.getContext().contentResolver, "enabled_notification_listeners")
                        val ok = listeners != null
                                && listeners.contains(mHost.getContext().packageName)
                        if (ok) {
                            AppLogger.i(TAG, "Notification listener enabled via shell ✓")
                            try {
                                Toast.makeText(mHost.getContext(),
                                        R.string.hud_notif_banner_toast_enabled,
                                        Toast.LENGTH_SHORT).show()
                            } catch (ignored: Throwable) {}
                            mCardHudNotif?.visibility = View.GONE
                            btnEnable?.isEnabled = true
                        } else {
                            AppLogger.w(TAG, "shell OK but listener not enabled — fallback. report=" + out)
                            openNotifListenerSettingsFallback()
                            btnEnable?.isEnabled = true
                        }
                    }
                }
                override fun onError(err: String?) {
                    mHost.runOnUiThread {
                        if (!mHost.isActivityAlive()) return@runOnUiThread
                        AppLogger.w(TAG, "one-click notif listener shell failed: " + err + " — fallback")
                        openNotifListenerSettingsFallback()
                        btnEnable?.isEnabled = true
                    }
                }
            })
        } catch (t: Throwable) {
            AppLogger.e(TAG, "enableHudNotifOneClick threw", t)
            openNotifListenerSettingsFallback()
            btnEnable?.isEnabled = true
        }
    }

    private fun openNotifListenerSettingsFallback() {
        try {
            val i = Intent(Settings.ACTION_NOTIFICATION_LISTENER_SETTINGS)
            i.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
            mHost.startActivity(i)
        } catch (t: Throwable) {
            AppLogger.w(TAG, "ACTION_NOTIFICATION_LISTENER_SETTINGS unavailable: " + t.message)
        }
    }

    companion object {
        private const val TAG = "PermBannerCoord"
    }
}
